### ISSUE TITLE
Search with MultiDirectionCollectionView

### ISSUE THREAD

**Issue by HPRaval** [kwandrews7/MultiDirectionCollectionView#3]
I want to add search functionality to this, But it gives me error on collectionview.reloadData(). It throws this exception

`Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'UICollectionView received layout attributes for a cell with an index path that does not exist: <NSIndexPath: 0x60000002fc40> {length = 2, path = 4 - 0}'`

To solve this i have changed layoutAttributesForElements to this

`override func layoutAttributesForElements(in rect: CGRect) -> [UICollectionViewLayoutAttributes]? {
        
        // Create an array to hold all elements found in our current view.
        var attributesInRect = [UICollectionViewLayoutAttributes]()
        
        // Check each element to see if it should be returned.
     
       for i in 0..<collectionView!.numberOfSections{
            for j in 0..<collectionView!.numberOfItems(inSection: i){
                let indexPath = IndexPath(item: j, section: i)
                if let attr = layoutAttributesForItem(at: indexPath) {
                    attributesInRect.append(attr)
                    
                }
            }
        }
        
        // Return list of elements.
        return attributesInRect
    }
`

But now it is creating problem in scrolling. After scroll data is miss placed. Will you please help me?

**Comment by kwandrews7** [comment]
Sometimes this is a little easier to debug with a complete sample I can
run. Could you upload your project to github.com so I could check it out
for you in the morning?

Thanks Kyle

**Comment by HPRaval** [comment]
I am sorry i cant, as it is quite big and i cant also you will be not able to access my server. My data is coming from local server.

well i have solved this error by this 

`override func invalidateLayout(with context: UICollectionViewLayoutInvalidationContext) {
        super.invalidateLayout(with: context)
        dataSourceDidUpdate = true
        cellAttrsDictionary.removeAll()
    }`

But now i am facing other issues. I have large set of data, when i scroll my data is miss placed

[mydata.zip](https://github.com/kwandrews7/MultiDirectionCollectionView/files/1060815/mydata.zip)

If you want i can upload sample code with sample data.

**Comment by kwandrews7** [comment]
@HPRaval your new issue is caused by the fact that you are basically causing your collection view to rebuild your entire data set when you filter. Updating the origin for thousands of cells is quite complex and can definitely cause performance issues. 

Sadly, I don't have a great easy answer for you. The best thing you can do at this point is likely hide the large processing hit from your users with by pushing the "filter" process (invalidating the layout and rebuilding all the attrs) to a background thread behind some sort of activity spinner overlay. This way you can show the user that something is happening without completely locking up your UI, and once the UI is finished updating, you can remove the overlay and they'll have access to the properly formatted table. 

Another approach I'd like to mention, it appears your data set doesn't really *need* to use a collection view. The main reason we used a collection view to solve our problem was because my data set could grow in to the n-th size in any direction. If your data set has a set horizontal width, you might be better off using an extra wide table view or a scroll view. This way you could basically filter entire rows and not have to update hundreds of individual cells.

**Comment by HPRaval** [comment]
Thnx for replay and your suggestion, well i have noticed that the problem of miss placing cell only occurs when i tried to put textfields in header for search. Without it, it is working smoothly. For now i have hard coded it, but i think i need to find other solution for this issue.

My code for header is

```
func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "Cell", for: indexPath) as! CustomCollectionViewCell
        

        if indexPath.section==0{ //header
            cell.backgroundColor = UIColor(colorLiteralRed: 47/255.0, green: 165/255.0, blue: 207/255.0, alpha: 1.0)
            cell.label.textColor = UIColor.white
             cell.label.text = data
        } else if indexPath.section==1{ //for search
            cell.label.isHidden = true
            cell.backgroundColor = UIColor.white
            let textField = UITextField(frame: cell.contentView.frame)
            textField.textAlignment = .center
            textField.tag = indexPath.item
            textField.font = textField.font?.withSize(11)
            textField.addTarget(self, action: #selector(searchItem(_:)), for: .editingChanged)
            
            if let search = search_text[indexPath.item]{
                textField.text = search
                textField.becomeFirstResponder()
            }
            cell.contentView.addSubview(textField)
            
        }else{ //for data
            cell.backgroundColor = UIColor.white
            cell.label.textColor = UIColor.black
             let data = String(describing: filtered[indexPath.section][indexPath.item])
            cell.label.text = data
        }
       
        //print("text....\(indexPath.section)....\(indexPath.item)....\(String(describing: cell.label.text))")
        return cell
    }


```
I have used collection view because my data comes from server, it is not fixed sized. Table is designed at server by admin, which can have n number of columns and rows. Thats why i am using this. 

Well i am also searching for any good alternative. If you have any then please let me know.
